### PR TITLE
New prophecy

### PR DESCRIPTION
**Pull request by juniuszhou**

### CODE REVIEW

**Comment by timlind** [comment]
@juniuszhou why were these changes made?

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: x/ethbridge/keeper/msg_server.go [hunk 1]
```diff
@@ -136,17 +137,19 @@ func (srv msgServer) Burn(goCtx context.Context, msg *types.MsgBurn) (*types.Msg
...
-	if status.Text == oracletypes.StatusText_STATUS_TEXT_SUCCESS {
-		if err = srv.Keeper.ProcessSuccessfulClaim(ctx, status.FinalClaim); err != nil {
+		if err != oracletypes.ErrProphecyFinalized {
+			logger.Error("bridge keeper failed to process claim.",
+				errorMessageKey, err.Error())
+			return nil, err
+		}
+
+	} else if status == oracletypes.StatusText_STATUS_TEXT_SUCCESS {
+		if err = srv.Keeper.ProcessSuccessfulClaim(ctx, msg.EthBridgeClaim); err != nil {
```
[the remark this answers is not in the document]
Comment (juniuszhou) [reply x/ethbridge/keeper/msg_server.go:152]: The err here is for keeper failed to process successful claim, the condition is like minting token not executed. 
The case for validator send claim to sifnode after prophecy already handled in the if branch.
the oracle keeper return a ErrProphecyFinalized error, we just ignore this error and not call ProcessSuccessfulClaim again. just emit events.
if err != nil {
		if err != oracletypes.ErrProphecyFinalized {
			logger.Error("bridge keeper failed to process claim.",
				errorMessageKey, err.Error())
			return nil, err
		}

File: x/ethbridge/alias.go [hunk 2]
```diff
@@ -23,22 +23,22 @@ const (
...
-	ErrJSONMarshalling                = types.ErrJSONMarshalling
-	NewEthereumAddress                = types.NewEthereumAddress
-	NewMsgCreateEthBridgeClaim        = types.NewMsgCreateEthBridgeClaim
-	MapOracleClaimsToEthBridgeClaims  = types.MapOracleClaimsToEthBridgeClaims
-	NewQueryEthProphecyParams         = types.NewQueryEthProphecyRequest
-	NewQueryEthProphecyResponse       = types.NewQueryEthProphecyResponse
+	NewKeeper         = keeper.NewKeeper
+	NewQuerier        = keeper.NewLegacyQuerier
+	NewEthBridgeClaim = types.NewEthBridgeClaim
+	// NewOracleClaimContent             = types.NewOracleClaimContent
```
Comment (ElliotFriedman) [inline x/ethbridge/alias.go:29]: Let's remove these commented out aliases.

File: x/ethbridge/alias.go [hunk 3]
```diff
@@ -47,9 +47,9 @@ var (
 )
 
 type (
-	Keeper                       = keeper.Keeper
-	EthBridgeClaim               = types.EthBridgeClaim //nolint:golint
-	OracleClaimContent           = types.OracleClaimContent
+	Keeper         = keeper.Keeper
+	EthBridgeClaim = types.EthBridgeClaim //nolint:golint
+	// OracleClaimContent           = types.OracleClaimContent
```
Comment (ElliotFriedman) [inline x/ethbridge/alias.go:52]: If this isn't used anymore, let's remove.

File: x/ethbridge/types/querier.go [hunk 4]
```diff
@@ -10,25 +10,22 @@ const (
...
-		TokenContractAddress:  tokenContractAddress.String(),
-		EthereumSender:        ethereumSender.String(),
+		ProphecyId: prophecyID,
 	}
 }
 
 // NewQueryEthProphecyResponse creates a new QueryEthProphecyResponse instance
-func NewQueryEthProphecyResponse(id string, status oracletypes.Status, claims []*EthBridgeClaim) QueryEthProphecyResponse {
+func NewQueryEthProphecyResponse(id []byte, status oracletypes.StatusText, claims []string) QueryEthProphecyResponse {
+	// claimValidators := []string{}
```
Comment (ElliotFriedman) [inline x/ethbridge/types/querier.go:21]: Remove commented out code assuming we don't need it.

File: x/ethbridge/types/test_common.go [hunk 5]
```diff
@@ -77,12 +78,13 @@ func CreateTestQueryEthProphecyResponse(t *testing.T, validatorAddress sdk.ValAd
 	testTokenAddress := NewEthereumAddress(TestTokenContractAddress)
 	ethBridgeClaim := CreateTestEthClaim(t, testContractAddress, testTokenAddress, validatorAddress,
 		testEthereumAddress, TestCoinsAmount, TestCoinsSymbol, claimType)
-	oracleClaim, _ := CreateOracleClaimFromEthClaim(ethBridgeClaim)
-	ethBridgeClaims := []*EthBridgeClaim{ethBridgeClaim}
+	// oracleClaim, _ := CreateOracleClaimFromEthClaim(ethBridgeClaim)
```
Comment (ElliotFriedman) [inline x/ethbridge/types/test_common.go:81]: Remove commented code.

File: x/oracle/keeper/keeper.go [hunk 6]
```diff
@@ -176,45 +154,25 @@ func (k Keeper) checkActiveValidator(ctx sdk.Context, validatorAddress sdk.ValAd
...
-		prophecy.Status.Text = types.StatusText_STATUS_TEXT_SUCCESS
-		prophecy.Status.FinalClaim = highestClaim
-	} else if highestPossibleConsensusRatio < k.consensusNeeded {
-		prophecy.Status.Text = types.StatusText_STATUS_TEXT_FAILED
-	}
+func (k Keeper) processCompletion(ctx sdk.Context, networkDescriptor types.NetworkDescriptor, prophecy types.Prophecy) types.Prophecy {
+	whiteList := k.GetOracleWhiteList(ctx, types.NewNetworkIdentity(networkDescriptor))
+	voteRate := whiteList.GetPowerRatio(prophecy.ClaimValidators)
 
+	if voteRate >= k.consensusNeeded {
```
Comment (ElliotFriedman) [inline x/oracle/keeper/keeper.go:173]: Looks good to me!

File: x/oracle/types/claim.go [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
 
 // NewClaim returns a new Claim
-func NewClaim(id string, validatorAddress string, content string) Claim {
-	return Claim{
-		Id:               id,
-		ValidatorAddress: validatorAddress,
-		Content:          content,
-	}
-}
+// func NewClaim(id string, validatorAddress string, content string) Claim {
```
Comment (ElliotFriedman) [inline x/oracle/types/claim.go:4]: If we don't use this code anymore, let's remove this file.

File: x/oracle/types/networkIdentity.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// IsValid check if the network id is valid
+func (n NetworkDescriptor) IsValid() bool {
+
+	if n == NetworkDescriptor_NETWORK_DESCRIPTOR_UNSPECIFIED {
+		return false
+	}
+
+	_, ok := NetworkDescriptor_name[int32(n)]
```
Comment (ElliotFriedman) [inline x/oracle/types/networkIdentity.go:50]: Where is `NetworkDescriptor_name` defined?

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: x/oracle/types/networkIdentity.go (hunk 8 above)
Comment (juniuszhou) [reply x/oracle/types/networkIdentity.go:50]: defined in network_descriptor.pb.go

File: x/oracle/types/claim.go (hunk 7 above)
Comment (juniuszhou) [reply x/oracle/types/claim.go:4]: accepted.

File: x/ethbridge/types/test_common.go (hunk 5 above)
Comment (juniuszhou) [reply x/ethbridge/types/test_common.go:81]: accepted.

File: x/ethbridge/types/querier.go (hunk 4 above)
Comment (juniuszhou) [reply x/ethbridge/types/querier.go:21]: accepted.

File: x/ethbridge/alias.go (hunk 3 above)
Comment (juniuszhou) [reply x/ethbridge/alias.go:52]: accepted.

File: x/ethbridge/alias.go (hunk 2 above)
Comment (juniuszhou) [reply x/ethbridge/alias.go:29]: accepted.

**Pull request merged**